File: app/src/main/java/com/imangazalievm/bubbble/presentation/mvp/userprofile/UserFollowersPresenter.java
package com.imangazalievm.bubbble.presentation.mvp.userprofile;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.imangazalievm.bubbble.domain.global.exceptions.NoNetworkException;
import com.imangazalievm.bubbble.domain.userprofile.UserFollowersInteractor;
import com.imangazalievm.bubbble.domain.global.models.Follow;
import com.imangazalievm.bubbble.domain.global.models.UserFollowersRequestParams;
import com.imangazalievm.bubbble.presentation.mvp.global.SchedulersProvider;
import com.imangazalievm.bubbble.presentation.utils.DebugUtils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

@InjectViewState
public class UserFollowersPresenter extends MvpPresenter<UserFollowersView> {

    private static final int PAGE_SIZE = 20;


    private UserFollowersInteractor userFollowersInteractor;
    private SchedulersProvider schedulersProvider;
    private long userId;
    private int currentMaxPage = 1;
    private List<Follow> followers = new ArrayList<>();
    private boolean isFollowersLoading = false;

    @Inject
    public UserFollowersPresenter(UserFollowersInteractor userFollowersInteractor,
                                  SchedulersProvider schedulersProvider,
                                  long userId) {
        this.userFollowersInteractor = userFollowersInteractor;
        this.schedulersProvider = schedulersProvider;
        this.userId = userId;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();

        getViewState().showFollowersLoadingProgress();
        loadMoreFollowers(currentMaxPage);
    }

    private void loadMoreFollowers(int page) {
        isFollowersLoading = true;
        UserFollowersRequestParams requestParams = new UserFollowersRequestParams(userId, page, PAGE_SIZE);
        userFollowersInteractor.getUserFollowers(requestParams)
                .observeOn(schedulersProvider.mainThread())
                .subscribe(this::onFollowersLoaded, this::onFollowersLoadError);
    }

    private void onFollowersLoaded(List<Follow> newFollowers) {
        isFollowersLoading = false;
        if (isFirstLoading()) {
            getViewState().hideFollowersLoadingProgress();
        } else {
            getViewState().hideFollowersLoadingMoreProgress();
        }

        followers.addAll(newFollowers);
        getViewState().showNewFollowers(newFollowers);
    }

    private void onFollowersLoadError(Throwable throwable) {
        isFollowersLoading = true;
        if (throwable instanceof NoNetworkException) {
            if (isFirstLoading()) {
                getViewState().hideFollowersLoadingProgress();
                getViewState().showNoNetworkLayout();
            } else {
                getViewState().hideFollowersLoadingMoreProgress();
                getViewState().showLoadMoreError();
            }
        } else {
            DebugUtils.showDebugErrorMessage(throwable);
        }
    }

    private boolean isFirstLoading() {
        return currentMaxPage == 1;
    }

    public void onLoadMoreFollowersRequest() {
        if (isFollowersLoading) {
            return;
        }

        getViewState().showFollowersLoadingMoreProgress();
        currentMaxPage++;
        loadMoreFollowers(currentMaxPage);
    }

    public void retryLoading() {
        if (isFirstLoading()) {
            getViewState().hideNoNetworkLayout();
            getViewState().showFollowersLoadingProgress();
        } else {
            getViewState().showFollowersLoadingMoreProgress();
        }

        loadMoreFollowers(currentMaxPage);
    }

    public void onFollowerClick(int position) {
        System.out.println(followers.size());
        System.out.println(followers.get(position));
        System.out.println(followers.get(position).getFollower());
        System.out.println(followers.get(position).getFollower().getId());
        getViewState().openUserDetailsScreen(followers.get(position).getFollower().getId());
    }


}

File: app/src/main/java/com/bubbble/presentation/userprofile/followers/UserFollowersPresenter.kt
package com.bubbble.presentation.userprofile.followers

import com.arellomobile.mvp.InjectViewState
import com.bubbble.core.exceptions.NoNetworkException
import com.bubbble.models.Follow
import com.bubbble.models.UserFollowersRequestParams
import com.bubbble.domain.userprofile.UserFollowersInteractor
import com.bubbble.presentation.global.mvp.BasePresenter
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import java.util.*

@InjectViewState
class UserFollowersPresenter @AssistedInject constructor(
    private val userFollowersInteractor: UserFollowersInteractor,
    @Assisted private val userId: Long
) : BasePresenter<UserFollowersView>() {

    private var currentMaxPage = 1
    private val followers: MutableList<com.bubbble.models.Follow> = ArrayList()
    private var isFollowersLoading = false

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.showFollowersLoadingProgress()
        loadMoreFollowers(currentMaxPage)
    }

    private fun loadMoreFollowers(page: Int) = launchSafe {
        isFollowersLoading = true
        val requestParams = com.bubbble.models.UserFollowersRequestParams(userId, page, PAGE_SIZE)
        try {
            val newFollowers = userFollowersInteractor.getUserFollowers(requestParams)
            followers.addAll(newFollowers)
            viewState.showNewFollowers(newFollowers)
        } catch (throwable: com.bubbble.core.exceptions.NoNetworkException) {
            if (isFirstLoading) {
                viewState.showNoNetworkLayout()
            } else {
                viewState.showLoadMoreError()
            }
        } finally {
            viewState.hideFollowersLoadingProgress()
            isFollowersLoading = false
        }
    }

    private val isFirstLoading: Boolean
        private get() = currentMaxPage == 1

    fun onLoadMoreFollowersRequest() {
        if (isFollowersLoading) {
            return
        }
        viewState.showFollowersLoadingMoreProgress()
        currentMaxPage++
        loadMoreFollowers(currentMaxPage)
    }

    fun retryLoading() {
        if (isFirstLoading) {
            viewState.hideNoNetworkLayout()
            viewState.showFollowersLoadingProgress()
        } else {
            viewState.showFollowersLoadingMoreProgress()
        }
        loadMoreFollowers(currentMaxPage)
    }

    fun onFollowerClick(position: Int) {
        println(followers.size)
        println(followers[position])
        println(followers[position].follower)
        println(followers[position].follower.id)
        viewState.openUserDetailsScreen(followers[position].follower.id)
    }

    companion object {
        private const val PAGE_SIZE = 20
    }

    @AssistedFactory
    interface Factory {
        fun create(userId: Long): UserFollowersPresenter
    }

}